R/QFeatures-class.R

```diff
@@ -1175,12 +1335,84 @@ setMethod("updateObject", "QFeatures",
 ##' @export
 dropEmptyAssays <- function(object, dims = 1:2) {
     stopifnot(inherits(object, "QFeatures"))
-    if (!all(dims %in% 1:2))
+    if (!all(dims %in% 1:2)) {
         stop("Argument 'dims' must be in '1:2'.")
-    if (1 %in% dims)
+    }
+    if (1 %in% dims) {
         object <- object[, , nrows(object) > 0]
-    if (2 %in% dims)
+    }
+    if (2 %in% dims) {
         object <- object[, , ncols(object) > 0]
-    if (!length(object)) return(QFeatures())
+    }
+    if (!length(object)) {
+        return(QFeatures())
+    }
     object
 }
+
+
+##' Set and Get QFeatures Type
+##'
+##' Developer-level functions to set and retrieve the type of a `QFeatures` 
+##' object. This type can help internal methods adapt their behavior to the
+##' structure of the data.
+##'
+##' @param object An instance of class [QFeatures].
+##' @param type `character(1)` defining the type of the QFeatures.
+##'     Must be one of the values returned by [validQFeaturesType()].
+##'
+##' @return 
+##' - `setQFeaturesType()`: returns the updated `QFeatures` object with
+##'   the type stored in its metadata.
+##' - `getQFeaturesType()`: returns a character string indicating the
+##'   type of the `QFeatures` object. If no type is explicitly set, 
+##'   it is inferred from the class of the experiments.
+##' - `validQFeaturesType()`: character vector of valid QFeatures types.
+##'
+##' @details
+##' These functions control an internal metadata field (`._type`) used to
+##' distinguish between different structural uses of `QFeatures` objects.
+##'
+##' @note These functions are intended for package developers and internal use.
+##'       End users should typically not call them directly.
+##'
+##' @rdname QFeatures-type
+##' @keywords internal
+##' @export
+setQFeaturesType <- function(object, type) {
+  stopifnot(inherits(object, "QFeatures"))
+  if (!type %in% validQFeaturesType()) {
+    stop(
+      "Invalid QFeatures type. Must be one of: ",
+      paste(validQFeaturesType(), collapse = ", ")
+    )
+  }
+  metadata(object)[["._type"]] <- type
+  object
+}
+
+##' @rdname QFeatures-type
+##' @keywords internal
+##' @export
+getQFeaturesType <- function(object) {
+  stopifnot(inherits(object, "QFeatures"))
+  type <- metadata(object)[["._type"]]
+  if (is.null(type)) {
+    warning(paste("No explicit type set for this QFeatures object,",
+                  "choosing a type in fonction of experiments classes"))
+    if (any(sapply(experiments(object),
+                   function(x) inherits(x, "SingleCellExperiment")))) {
+      type <- "scp"
+    } else {
+      type <- "bulk"
+    }
+  }
+  type
+}
+
+##' @rdname QFeatures-type
+##' @keywords internal
+##' @export
+validQFeaturesType <- function() {
```

Comment:
I think this function should be called `validQFeaturesTypes()`. See also comment in PR #236.